crates/node/builder/src/builder/states.rs

```diff
@@ -149,66 +160,188 @@ impl<T: FullNodeTypes, C: NodeComponents<T>> Clone for NodeAdapter<T, C> {
 /// A fully type configured node builder.
 ///
 /// Supports adding additional addons to the node.
-pub struct NodeBuilderWithComponents<
-    T: FullNodeTypes,
-    CB: NodeComponentsBuilder<T>,
-    AO: NodeAddOns<NodeAdapter<T, CB::Components>>,
-> {
+pub struct NodeBuilderWithComponents<T: BuilderInternals> {
     /// All settings for how the node should be configured.
-    pub config: NodeConfig<<T::Types as NodeTypes>::ChainSpec>,
-    /// Adapter for the underlying node types and database
-    pub adapter: NodeTypesAdapter<T>,
-    /// container for type specific components
-    pub components_builder: CB,
-    /// Additional node extensions.
-    pub add_ons: AddOns<NodeAdapter<T, CB::Components>, AO>,
+    pub adapter: T,
```

Comment:
I think what we should try here is introducing an actual Adapter struct on this level, which I should would allow us to hide most of this, this is like nodeadapter but over internals.

then this maybe doesn't cascade